### PR TITLE
spirv-fuzz: Fix OpPhi handling in DuplicateRegionWithSelection

### LINKED ISSUE #3787: spirv-fuzz: Apply transformation for blocks with pointers (and other types) when adding OpPhi is unnecessary.
Adding OpPhi for instructions with pointer types requires CapabilityVariablePointers in a shader. Currently the transformation adds OpPhi instructions to all instructions from original region which are available at the end of the region. If the shader does not have a CapabilityVariablePointers, this limits the applicability of the transformation. Consider not adding OpPhi instructions corresponding to pointers which are not used after the region, so that the transformation is still applicable to regions contaning them.

### PR DESCRIPTION
**Pull request by afd**
Avoid generating OpPhi on void types, and allow the transformation to
take place on regions that produce pointer and sampled image result
ids if such ids are not used after the region.

Fixes #3787.

### CODE REVIEW

File: source/fuzz/transformation_duplicate_region_with_selection.cpp [hunk 1]
```diff
@@ -237,43 +237,48 @@ bool TransformationDuplicateRegionWithSelection::IsApplicable(
...
+      // this not just for instructions that are already used after the region,
+      // but for all instructions so that the phi is available to future
+      // transformations.
+      if (AvailableAfterRegion(instr, exit_block, ir_context)) {
+        if (!ValidOpPhiArgument(instr, ir_context)) {
+          // The instruction cannot be used as an OpPhi argument.  This is a
+          // blocker if there are uses of the instruction after the region.
+          // Otherwise we can simply avoid generating an OpPhi for this
+          // instruction and its duplicate.
+          if (!ir_context->get_def_use_mgr()->WhileEachUse(
```
Comment (Vasniktel) [inline source/fuzz/transformation_duplicate_region_with_selection.cpp:253]: Can we use `WhileEachUser` here?

**Review by Vasniktel — COMMENTED** [review]
Minor comments.

File: source/fuzz/transformation_duplicate_region_with_selection.cpp [hunk 2]
```diff
@@ -237,43 +237,48 @@ bool TransformationDuplicateRegionWithSelection::IsApplicable(
...
-
-        // Every instruction with a result id available at the end of the region
-        // must be present in the map |original_id_to_phi_id|, unless overflow
-        // ids are present.
-        if (original_id_to_phi_id.count(instr.result_id()) == 0) {
-          if (!transformation_context.GetOverflowIdSource()->HasOverflowIds()) {
+      // If the instruction is available at the end of the region then we would
+      // like to be able to add an OpPhi instruction at the merge point of the
+      // duplicated region to capture the values computed by both duplicates of
+      // the instruction, so that this is available to after the region.  We do
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/transformation_duplicate_region_with_selection.cpp:243]: Fixed, thanks.

File: source/fuzz/transformation_duplicate_region_with_selection.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/transformation_duplicate_region_with_selection.cpp:253]: Good point - done.

File: source/fuzz/transformation_duplicate_region_with_selection.cpp [hunk 3]
```diff
@@ -484,26 +489,31 @@ void TransformationDuplicateRegionWithSelection::Apply(
...
-          (&instr == &*exit_block->tail() ||
-           fuzzerutil::IdIsAvailableBeforeInstruction(
-               ir_context, &*exit_block->tail(), instr.result_id()))) {
-        // Add the OpPhi instruction for every result id that is
-        // available at the end of the region (the last instruction
-        // of the |exit_block|)
+      if (instr.result_id() == 0) {
+        continue;
+      }
+      if (ir_context->get_def_use_mgr()->GetDef(instr.type_id())->opcode() ==
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/transformation_duplicate_region_with_selection.cpp]: You're right - good catch.  I didn't initially check for the OpTypeVoid case in that function and must have forgotten to eliminate this check when I started doing so.

**Review by Vasniktel — APPROVED** [review]

**Pull request merged**